src/pybind/mgr/cephadm/upgrade.py

```diff
@@ -610,27 +616,43 @@ def _prepare_for_mds_upgrade(
 
             # scale down this filesystem?
             if mdsmap["max_mds"] > 1:
-                self.mgr.log.info('Upgrade: Scaling down filesystem %s' % (
-                    fs_name
-                ))
-                if fscid not in self.upgrade_state.fs_original_max_mds:
-                    self.upgrade_state.fs_original_max_mds[fscid] = mdsmap['max_mds']
-                    self._save_upgrade_state()
-                ret, out, err = self.mgr.check_mon_command({
-                    'prefix': 'fs set',
-                    'fs_name': fs_name,
-                    'var': 'max_mds',
-                    'val': '1',
-                })
-                continue_upgrade = False
-                continue
+                if self.upgrade_state.fail_fs and (mdsmap['flags'] & 1) and \
+                        (len(mdsmap['up']) > 0):
+                    # only fail fs if fail_fs = true and fs is joinable and
+                    # there does exist mds in mdsmap['up']
+                    self.mgr.log.info(f'Upgrade: failing fs {fs_name} for '
+                                      f'rapid multi-rank mds upgrade')
+                    ret, out, err = self.mgr.check_mon_command({
+                        'prefix': 'fs fail',
+                        'fs_name': fs_name
+                    })
+                    continue_upgrade = True
```

Comment:
is it expected that the `fs fail` command will work basically immediately? With the scale down we use that `continue_upgrade` var to have us essentially loop until we see that scale down is complete before we continue. Do we need something similar here where we say we can't continue until the fs is not joinable and no mds daemons are up or is it okay to just run `fail fs` then immediately start upgrading mds daemons?

Reply:
Thanks for bringing it up. I don't have much idea about how "immediate" it is. I'll look into it and update this code accordingly.

Reply:
@adk3798 as per [this](https://docs.ceph.com/en/latest/cephfs/administration/#taking-the-cluster-down-rapidly-for-deletion-or-disaster-recovery) official doc, it says "To allow rapidly deleting a file system (for testing) or to quickly bring the file system and MDS daemons down, use the `fs fail` command".